src/eva/vision/data/wsi/patching/coordinates.py

```diff
@@ -55,8 +58,28 @@ def from_file(
         mpp_ratio = target_mpp / level_mpp
         scaled_width, scaled_height = int(mpp_ratio * width), int(mpp_ratio * height)
 
-        for x, y in sampler.sample(scaled_width, scaled_height, wsi.level_dimensions[level_idx]):
-            x_y.append((x, y))
+        if not isinstance(sampler, samplers.ForegroundGridSampler):
```

Comment:
I'd replace this by something like this:

```
if isinstance(sampler, samplers.ForegroundSampler):
    mask = get_mask(...)  # do the scaling in this function as well

    for x, y in sampler.sample(
        scaled_width, scaled_height, wsi.level_dimensions[level_idx], mask=mask
     ):
        x_y.append((x, y))
```

We can add `*args` to the init args of the samplers, so passing `mask=mask` argument won't do anything, or we can build a dict `sample_args` here and call `sample(*sample_args)` , and add the `mask` key only for foreground samplers. Maybe the latter is cleaner?

Then in `samplers.py` we can add an abstract base class for foreground samplers (as eventually we might have multiple of those)

```
class ForegroundSampler(abc.ABC, Sampler):
    """Base class for samplers with foreground filtering capabilities."""

    @abc.abstractmethod
    def sample(
        self,
        width: int,
        height: int,
        layer_shape: tuple[int, int],
        mask: np.ndarray,
    ) -> Generator[Tuple[int, int], None, None]:
        """Iterator that samples patches."""

    @abc.abstractmethod
    def is_foreground(...):
        pass
```

Reply:
refactored accordingly. However, note that now the ForegroundGridSampler doesn't inherit from GridSampler (due to issue with signature when inheriting from both, ForegroundSampler and GridSampler). Moved the shared grid-sampling logic to the function `_get_grid_coords_and_indices` instead

Reply:
I see, I think it looks good now, thanks for addressing this. 🚀